src/main/java/build/buildfarm/instance/stub/StubInstance.java

```diff
@@ -240,23 +251,90 @@ public void write(byte[] b, int off, int len) throws IOException {
 
   @Override
   public InputStream newStreamInput(String name) {
-    throw new UnsupportedOperationException();
+    Iterator<ReadResponse> replies = bsBlockingStub
+      .get()
+      .read(ReadRequest.newBuilder().setResourceName(name).build());
+    return new InputStream() {
+      InputStream chunkIn = nullInputStream;
+      boolean closed = false;
+
+      @Override
+      public int available() throws IOException {
+        if (closed) {
+          throw new IOException("stream is closed");
+        }
+        if (chunkIn.available() == 0) {
+          advance();
+        }
+        return chunkIn.available();
+      }
+
+      @Override
+      public int read() throws IOException {
+        if (closed) {
+          throw new IOException("stream is closed");
+        }
+        byte[] b = new byte[1];
+        if (read(b) == -1) {
+          return -1;
+        }
+        return b[0];
+      }
+
+      @Override
+      public int read(byte[] b, int off, int len) throws IOException {
+        if (closed) {
+          throw new IOException("stream is closed");
+        }
+        int totalLen = 0;
+        while (len != 0) {
+          int readLen = chunkIn.read(b, off, len);
+          if (readLen == -1) {
+            advance();
+            readLen = chunkIn.read(b, off, len);
+            if (readLen == -1) {
+              return totalLen == 0 ? -1 : totalLen;
```

Comment:
I think it would be better to have an end-of-stream boolean indicator and use that to detect end of stream. Then you can just continue after the call to advance().